BizHawk.Client.Common/movie/tasproj/StateManagerDecay.cs

```diff
@@ -59,152 +59,149 @@ public StateManagerDecay(TasStateManager tsm)
 			_align = false;
 		}
 
-		// todo: go through all states once, remove as many as we need. refactor to not need goto
+		// todo: go through all states once, remove as many as we need. ~~~refactor to not need goto~~~ yw -Yoshi
 		public void Trigger(int decayStates)
 		{
-			for (; decayStates > 0 && _tsm.StateCount > 1;)
+			while (decayStates > 0 && _tsm.StateCount > 1) Trigger_Sub(decayStates);

```

Comment:
Does this need to be `while (...) decayStates = Trigger_Sub(decayStates);` (with the returns and fallthrough in `Trigger_Sub` returning the modified value of `decayStates`)?

Reply:
Can't tell without testing. This logic is quite complex and it can break in ways that, when fixed, lead to new bugs recursively. This function is so influential and fragile that I'd even move any refactors of it to yet another PR. Or just give up until we hit new bugs in it, then we'd just **have** to refactor it. Let's not touch it until I have time to meditate on this once again.

Reply:
What I meant was, in C#, method int params are mutable, but do changes affect the copy passed to the function?

Reply:
Which function? I may only know what was happening in the old code.

Reply:
HEAD / sans goto, I've added the _Sub function and passed decayStates to it (I've done this elsewhere too), but the sub function calls decayStates--.

Reply:
Not resolved until I can take a proper closer look.

Reply:
I have since learned what value types are and int is one of them, so yes.